### ISSUE TITLE
Slow async UDF

### ISSUE THREAD

**Issue by dolezj24** [Excel-DNA/ExcelDna#232]
Hello, while using Excel DNA async UDFs we experienced some performance issues. The main problem seems to be when you are using UDF directly in the cell and with larger quantity. There is a simple sample of UDF we tested on:
`
```
[ExcelFunction(Description = "Validate", IsThreadSafe = true)]
public static void Validate(object value, ExcelAsyncHandle asyncHandle)
{
    asyncHandle.SetResult(1);
}
````
When you create sheet with 100k values of `=Validate("dummy")` without addin added to Excel and then you reopen sheet with addin activated, it takes about 30s to calculate. Also, there seems to be theshold around 25k values, because sheet with 20k values is calculated almost immediately (<1s). So steps to replacate issue are:

1. Create sheet with 100k values of `=Validate("dummy")`

**Comment by govert** [comment]
One helpful comparison would be to make a native add-in in C based on the XLL SDK with a similar async function, and compare the performance. While the C# add-in has some overhead, we expect that to by fairly constant.

One gotcha regarding async functions in C# (which does not seem directly applicable to the issue reported here) is that the .NET ThreadPool ramps up rather slowly - so if your async tasks are using up many threads you need to either expand the ThreadPool size, or at least take some care to understand how the defulat ThreadPool behaves. That doesn't seem like the issue here though.

**Comment by mvorisek** [comment]
Issue submited to Microsoft via answers.microsoft.com: https://answers.microsoft.com/en-us/msoffice/forum/msoffice_excel-mso_other-mso_2016/excel-core-issue-slow-async-udf/dd9c47a0-02eb-4d69-982e-4bebf9edac6d

**Comment by OlegRa** [comment]
I've made similar testing using pure C/C++ implementation (just slightly modified XLL SDK example project) and it demonstrates similar issue on my Office 365 version (monthly updates channel). I'm not sure about old versions of Excel but this problem exists at least in latest builds.
Most interesting aspect of this problem - progress is slowing down in time. So looks like first xlAsyncReturn calls run in time, comparable with non-thread-safe version of same UDF and execution time increased a little bit after each call.

**Comment by govert** [comment]
This discussion on doing the async update back to Excel in batches might be relevant:
https://groups.google.com/forum/#!topic/exceldna/NZLGGZFv-fM

If it doesn't work as-is with the code I posted, it should be still be easy to make work.

**Comment by mvorisek** [comment]
@OlegRa Yes, we observed this roughly in the status bar. The first percents are much faster than to the end.

@govert I think this is unrelated. Our tests show that the described issue here, i.e. (about) exponential complexity (growing linearly per UDF), can be observed even if no result is set (the time before the Excel will invoke all UDFs is about the same no matter if `SetResult` or `SetException` are called immediately or actually never called).
